modules/library/metadata/src/main/java/org/geotools/measure/WktUnitFormatterFactory.java

```diff
@@ -0,0 +1,137 @@
+/*
+ *    GeoTools - The Open Source Java GIS Toolkit
+ *    http://geotools.org
+ *
+ *    (C) 2021, Open Source Geospatial Foundation (OSGeo)
+ *
+ *    This library is free software; you can redistribute it and/or
+ *    modify it under the terms of the GNU Lesser General Public
+ *    License as published by the Free Software Foundation;
+ *    version 2.1 of the License.
+ *
+ *    This library is distributed in the hope that it will be useful,
+ *    but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ *    Lesser General Public License for more details.
+ *
+ *    This package contains documentation from OpenGIS specifications.
+ *    OpenGIS consortium's work is fully acknowledged here.
+ */
+package org.geotools.measure;
+
+import static org.geotools.measure.SimpleUnitFormatForwarder.BaseUnitFormatter;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import javax.measure.Quantity;
+import javax.measure.Unit;
+import tech.units.indriya.unit.TransformedUnit;
+
+/**
+ * UnitFormat configured to parse units. Since usually we don't know the citation in use for a
+ * particular unit literal definition, this format includes the aliases of EPSG and ESRI citations,
+ * in order to be able to parse the widest possible range of units.
+ */
+public final class WktUnitFormatterFactory {
+
+    public static UnitFormatter getUnitFormatterSingleton() {
+        return INSTANCE;
+    }
+
+    public static UnitFormatter create() {
+        return new WktUnitFormatter(UNIT_DEFINITIONS);
+    }
+
+    private static final List<UnitDefinition> UNIT_DEFINITIONS =
+            Stream.of(
+                            UnitDefinitions.DIMENSIONLESS,
+                            UnitDefinitions.CONSTANTS,
+                            UnitDefinitions.SI_BASE,
+                            UnitDefinitions.SI_DERIVED,
+                            UnitDefinitions.NON_SI,
+                            UnitDefinitions.US_CUSTOMARY,
+                            UnitDefinitions.WKT)
+                    .flatMap(Collection::stream)
+                    .collect(Collectors.toList());
+
+    private static final WktUnitFormatter INSTANCE = (WktUnitFormatter) create();
+
+    private WktUnitFormatterFactory() {}
+
+    static class WktUnitFormatter extends BaseUnitFormatter implements UnitFormatter {
+
+        private HashMap<UnitWrapper, Unit<?>> unitWrapperToUnitMap;
+
+        WktUnitFormatter(List<UnitDefinition> unitDefinitions) {
+            super(unitDefinitions);
+        }
+
+        @Override
+        protected void addUnit(Unit<?> unit) {
+            if (unitWrapperToUnitMap == null) {
+                unitWrapperToUnitMap = new HashMap<>();
+            }
+            unitWrapperToUnitMap.put(new UnitWrapper(unit), unit);
```

Comment:
@jodygarnett There is probably a simpler way to achieve this now, if we assume that the instances we build are immutable:

Instead of subclassing `BaseUnitFormatter` and implementing the `addUnit` hook method, it would probably suffice to build `unitWrapperToUnitMap` in the `WktUnitFormatter` constructor:

    unitWrapperToUnitMap = unitDefinitions.stream()
        .map(UnitDefinition::getUnit)
        .collect(Collectors.toMap(UnitWrapper::new, Function.identity()));

Reply:
I have implemented this, and subsequently created the possibility to enforce that unit formatters aren't mutated after construction.